src/api/routes/rosetta/construction.ts

```diff
@@ -297,5 +324,174 @@ export function createRosettaConstructionRouter(db: DataStore): RouterWithAsync
     }
   });
 
+  //construction/payloads endpoint
+  router.postAsync('/payloads', async (req, res) => {
+    const valid: ValidSchema = await rosettaValidateRequest(req.originalUrl, req.body);
+    if (!valid.valid) {
+      res.status(400).json(makeRosettaError(valid));
+      return;
+    }
+
+    const options = getOptionsFromOperations(req.body.operations);
+    if (options == null) {
+      res.status(400).json(RosettaErrors.invalidOperation);
+      return;
+    }
+
+    const amount = options.amount;
+    if (!amount) {
+      res.status(400).json(RosettaErrors.invalidAmount);
+      return;
+    }
+
+    const fees = options.fee;
+    if (!fees) {
+      res.status(400).json(RosettaErrors.invalidFees);
+      return;
+    }
+
+    const publicKeys: RosettaPublicKey[] = req.body.public_keys;
+    if (!publicKeys) {
+      res.status(400).json(RosettaErrors.emptyPublicKey);
+      return;
+    }
+    for (const key of publicKeys) {
+      if (key.curve_type !== 'secp256k1') {
+        res.status(400).json(RosettaErrors.invalidCurveType);
+        return;
+      }
+    }
+
+    const recipientAddress = options.token_transfer_recipient_address;
+    if (!recipientAddress) {
+      res.status(400).json(RosettaErrors.invalidRecipient);
+      return;
+    }
+    const senderAddress = options.sender_address;
+
+    if (!senderAddress) {
+      res.status(400).json(RosettaErrors.invalidSender);
+      return;
+    }
+
+    const accountInfo = await new StacksCoreRpcClient().getAccount(senderAddress);
+
+    let tokenTransferOptions: UnsignedTokenTransferOptions | UnsignedMultiSigTokenTransferOptions;
+
+    if (publicKeys.length > 1) {
+      //multi signature
+      const publicKeysStrings = publicKeys.map(key => {
+        return key.hex_bytes;
+      });
+      tokenTransferOptions = {
+        recipient: recipientAddress,
+        amount: new BN(amount),
+        fee: new BN(fees),
+        publicKeys: publicKeysStrings,
+        numSignatures: publicKeys.length,
```

Comment:
did you intend for `publicKeys.length` to be the value here? This will mean that all of the keys associated with `publicKeys` will have to eventually sign this tx...

typically the creator of a multi-sig transaction is able to specify what this value should be.

Reply:
I'm assuming the caller is only specifying keys whose signatures are required. It means the caller is the creator of mulit-sig/single-sig transaction here. I can not find any other param in rosetta which will indicate signatures and public keys can be different.

Reply:
@asimm241 can you verify that other Rosetta implementations do not support m-of-n multisig transactions?

Reply:
@zone117x I haven't found any implementation to support multi-sig for rosetta in other blockchains. I have found this question https://community.rosetta-api.org/t/construction-api-and-multi-sig-accounts/208. This states that they are targeting m-of-n multi-sig for rosetta `v1.5.0`

Keeping this in view should we postpone the support of multi-sig. 

Moreover, BTW rosetta docs do not require public keys in payloads request but `makeUnsignedTokenTransferTransaction` requires a public key(s), is there any other way to create a transaction without specifying the public key(s)?

Reply:
> This states that they are targeting m-of-n multi-sig for rosetta v1.5.0. Keeping this in view should we postpone the support of multi-sig.

We can postpone m-of-n until then 👍 

> rosetta docs do not require public keys in payloads request but makeUnsignedTokenTransferTransaction requires a public key(s)

It appears to be not-required to allow support for blockchains that do not need public keys. Is there anywhere that states the endpoint must work when public keys are not provided?

Reply:
No, I haven't encountered anything like that so far, I'm searching for it, I'm also thinking of asking a question on their community forum. Let's see what happens.